pkg/csi/hostpathimpl.go

```diff
@@ -0,0 +1,271 @@
+		return fmt.Errorf("mountHostPathVolume: fail to mount %s to %s: %s", sourcePath, targetPath, err.Error())
+	}
+	return nil
+}
+
+func (ns *hostPathNsImpl) setProjectQuota(ctx context.Context, basePath string, volumeHostPath string, capacity int64) error {
+	if notMountPoint, err := ns.common.k8smounter.IsLikelyNotMountPoint(basePath); err != nil {
+		return fmt.Errorf("setProjectQuota: failed to check if basePath %s is mounted, err: %s", basePath, err.Error())
+	} else if notMountPoint {
+		return fmt.Errorf("setProjectQuota: basePath %s is not a mount point, unable to set quota", basePath)
+	}
+
+	cmd := `
+set -ex
+# fs stores the file system of mount
+FS=$(stat -f -c %T "{{ .BasePath }}")
+# check if fs is xfs or ext4 (output of stat is ext2/ext3)
+# PID is the last project Id in the directory
+# xfs_quota project(xfs) or chattr +P (ext4) initializes project with new project id
+# xfs_quota limit(xfs) or repquota (ext4) sets the quota according to limits defined
+if [[ "$FS" == "xfs" ]]; then
+  PID=$(xfs_quota -x -c 'report -h' "{{ .BasePath }}" | tail -2 | awk 'NR==1{print substr ($1,2)}+0')
+  PID=$(expr $PID + 1)
+  xfs_quota -x -c 'project -s -p  {{ .ProjectPath }}' $PID "{{ .BasePath }}"
+  xfs_quota -x -c 'limit -p bsoft={{ .SoftLimit }} bhard={{ .HardLimit }}' $PID "{{ .BasePath }}"
+elif [[ "$FS" == "ext2/ext3" ]]; then
+  PID=$(repquota -P "{{ .BasePath }}" | tail -3 | awk 'NR==1{print substr ($1,2)}+0')
+  PID=$(expr $PID + 1)
+  chattr +P -p $PID "{{ .ProjectPath }}"
+  setquota -P $PID {{ .UpperSoftLimit }} {{ .UpperHardLimit }} 0 0 "{{ .BasePath }}"
+fi
+    `
+	tmpl, err := template.New("").Parse(cmd)
+	if err != nil {
+		return fmt.Errorf("failed to parse template, err: %w", err)
+	}
+	capacityKB := (capacity + 1023) / 1024
+	limit := fmt.Sprintf("%dk", capacityKB)
+	var strBuilder strings.Builder
+	if err = tmpl.Execute(&strBuilder, struct {
+		BasePath                  string
+		ProjectPath               string
+		SoftLimit, UpperSoftLimit string
+		HardLimit, UpperHardLimit string
+	}{
+		BasePath:       basePath,
+		ProjectPath:    volumeHostPath,
+		SoftLimit:      limit,
+		HardLimit:      limit,
+		UpperSoftLimit: strings.ToUpper(limit),
+		UpperHardLimit: strings.ToUpper(limit),
+	}); err != nil {
+		return fmt.Errorf("failed to render template, err: %w", err)
+	}
+
+	const (
+		lockTimeout  = "30" // seconds
+		lockFileName = "setquota.lock"
+	)
+	lockfile := filepath.Join(basePath, lockFileName)
+	out, err := ns.common.osTool.RunCommand("flock", []string{
+		"-w", lockTimeout, "-x", lockfile, "-c", strBuilder.String(),
+	})
+	if err != nil {
+		err = fmt.Errorf("setProjectQuota failed, error: %w, output: %s", err, out)
+	}
+	log.Infof("setProjectQuota: cmd output: %s, err: %v", out, err)
+	return err
+}
+
+func getVerifiedHostPath(volumeID string, params map[string]string) (string, string, error) {
+	basePath := params[hostPathScTag]
```

Comment:
Maybe it's better to move validation check of arguments to CreateVolume stage? Expose error as quick as possible.

Reply:
It definitely should be, I missed it. Done.